core/src/com/unciv/logic/automation/unit/EspionageAutomation.kt

```
@@ -1,33 +1,83 @@
+            (!it.isDoingWork() || (it.action == SpyAction.CounterIntelligence && spiesDoingCounterIntelligence > 2))}
+        for (spy in spiesToMove) {
+            val randomSeed = spies.size + spies.indexOf(spy) + civInfo.gameInfo.turns
+            val randomAction = Random(randomSeed).nextInt(10)
+
+            // Try each operation based on the random value and the success rate
+            // If an operation was not successfull try the next one
+            if (randomAction <= 7 && automateSpyStealTech(spy)) {
                 continue
+            } else if (randomAction <= 9 && automateSpyRigElection(spy)) {
+                continue
+            } else if (automateSpyCounterInteligence(spy)) {
+                continue
+            } else if (spy.isDoingWork()) {
+                continue // We might have been doing counter intelligence and wanted to look for something better
+            } else {
+                // Retry all of the operations one more time
+                if (automateSpyStealTech(spy)) continue
+                if (automateSpyRigElection(spy)) continue
+                if(automateSpyCounterInteligence(spy)) continue
             }
-            if (spy.action == SpyAction.None) {
-                spy.moveTo(civInfo.getKnownCivs().filter { otherCiv -> otherCiv.isMajorCiv() && otherCiv.cities.any { it.getCenterTile().isVisible(civInfo) }}
-                    .toList().randomOrNull()?.cities?.filter { it.getCenterTile().isVisible(civInfo) }?.randomOrNull())
-            }
+            // There is nothing for our spy to do, put it in a random city
+            spy.moveTo(civInfo.gameInfo.getCities().filter { it.getCenterTile().isVisible(civInfo) && spy.canMoveTo(it) }.toList().randomOrNull())
         }
     }
+
+    /**
+     * Moves the spy to a city that we can steal a tech from
+     */
+    fun automateSpyStealTech(spy: Spy): Boolean {
+        if (civsToStealFrom.isNotEmpty()) {
```

Fail faster - switch the if to "if X.empty() return false" for zero nesting